File: trainer/frontend/src/sim/physics.ts
/**
 * TypeScript port of trainer/backend/physics.py — same algorithm, same
 * math, kept in sync by hand (there's no shared source between the two
 * languages). See physics.py's module docstring for the full rationale
 * of the two-pass Jacobi solver; this file only restates the mechanics
 * needed to reproduce it exactly.
 *
 * The constants themselves are *not* hand-copied, unlike the algorithm —
 * they arrive at runtime as a PhysicsConfig, broadcast by train_server.py
 * on every generation message (see net/trainingSocket.ts), sourced
 * directly from physics.py's own module constants. That's the single
 * source of truth this file used to duplicate as its own hardcoded
 * RADIUS/TENSION_STIFFNESS/etc. exports — tuning a constant backend-side
 * used to require a matching hand-edit here to avoid silent drift; now
 * there's nothing here to edit.
 *
 * The one structural difference from physics.py: relaxSteps() is a
 * generator that yields the in-progress positions after every single
 * correction (settle pass, then cleanup pass) instead of only returning
 * the converged result — driving this generator a few steps at a time
 * from requestAnimationFrame is what makes the settling motion animate
 * smoothly instead of jumping straight to rest.
 *
 * Performance: TENSION_RANGE is a small, fixed interaction radius, so in
 * any reasonably spread-out graph most pairs of nodes are never within
 * range of each other — the original implementation still checked every
 * one of them, every iteration, via a dense O(n^2) double loop. This
 * version buckets nodes into a uniform grid sized to the query radius
 * (mirroring physics.py's KD-tree query_pairs, minus the library — JS
 * has no builtin spatial tree) and only checks the cell a node is in
 * plus its neighbors, which is where essentially all the frame time was
 * going once a replay reached a couple hundred nodes.
 */

/**
 * Everything physics.py hardcodes as module constants, sent over the
 * wire instead of mirrored here — see this file's own docstring. Field
 * names match train_server.py's PHYSICS_CONFIG dict keys exactly (plain
 * hand-matched, same as every other server->client message shape in
 * this codebase — see e.g. ReplayConfig in runner.ts).
 */
export interface PhysicsConfig {
  radius: number
  contactDistance: number
  tensionRange: number
  tensionStiffness: number
  settleStiffness: number
  settleIterations: number
  cleanupIterations: number
  settleConvergenceTol: number
  cleanupConvergenceTol: number
  // False removes the hard "circles never overlap" guarantee entirely —
  // mirrors physics.py's ENABLE_COLLISION, see that constant's own
  // comment for the full rationale.
  collisionEnabled: boolean
}

export type Vec2 = [number, number]
type Pair = [number, number]

// The 5 directions that, applied to every occupied cell, cover the full
// 3x3 Moore neighborhood of cell-pairs exactly once each — no pair
// double-counted, none missed. (0,0) is the cell itself (self-pairs);
// the other 4 are the "forward" half of the 8 neighbors — whichever
// cell would be their "backward" counterpart picks them up when *it's*
// the reference cell. Verified against brute force before shipping.
const NEIGHBOR_OFFSETS: Pair[] = [
  [0, 0],
  [1, 0],
  [0, 1],
  [1, 1],
  [-1, 1],
]

function findPairs(pos: Vec2[], radius: number): Pair[] {
  const grid = new Map<string, number[]>()
  for (let i = 0; i < pos.length; i++) {
    const cx = Math.floor(pos[i][0] / radius)
    const cy = Math.floor(pos[i][1] / radius)
    const key = `${cx},${cy}`
    const bucket = grid.get(key)
    if (bucket) bucket.push(i)
    else grid.set(key, [i])
  }

  const radiusSq = radius * radius
  const pairs: Pair[] = []

  for (const [key, bucket] of grid) {
    const [cxStr, cyStr] = key.split(",")
    const cx = Number(cxStr)
    const cy = Number(cyStr)
    for (const [dx, dy] of NEIGHBOR_OFFSETS) {
      const neighborBucket = grid.get(`${cx + dx},${cy + dy}`)
      if (!neighborBucket) continue
      const sameCell = dx === 0 && dy === 0
      for (let a = 0; a < bucket.length; a++) {
        const startB = sameCell ? a + 1 : 0
        for (let b = startB; b < neighborBucket.length; b++) {
          const i = bucket[a]
          const j = neighborBucket[b]
          const ddx = pos[i][0] - pos[j][0]
          const ddy = pos[i][1] - pos[j][1]
          if (ddx * ddx + ddy * ddy <= radiusSq) pairs.push([i, j])
        }
      }
    }
  }
  return pairs
}

function tensionCompatibility(idVectors: number[][]): number[][] {
  const n = idVectors.length
  const norms = idVectors.map((v) => Math.hypot(...v))
  const unit = idVectors.map((v, i) => {
    const norm = norms[i] < 1e-9 ? 1.0 : norms[i]
    return v.map((c) => c / norm)
  })
  const compat: number[][] = Array.from({ length: n }, () =>
    new Array(n).fill(0)
  )
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      let dot = 0
      for (let k = 0; k < unit[i].length; k++) dot += unit[i][k] * unit[j][k]
      compat[i][j] = Math.min(1, Math.max(0, dot))
    }
  }
  return compat
}

/**
 * Jacobi-averaged correction from a sparse set of candidate pairs (found
 * within some search radius — see findPairs). `stiffnessOf` returning
 * null means "not actually active" (e.g. a settle-phase candidate that
 * turned out to be past tensionRange at the queried radius's boundary,
 * or a cleanup-phase candidate that isn't really colliding) — such a
 * pair contributes nothing and isn't counted, exactly like the old dense
 * version's active_mask gating both magnitude and count together.
 */
function pairCorrections(
  n: number,
  pos: Vec2[],
  pairs: Pair[],
  target: number,
  stiffnessOf: (i: number, j: number, d: number) => number | null,
  freeMask: boolean[]
): Vec2[] {
  const sumX = new Array(n).fill(0)
  const sumY = new Array(n).fill(0)
  const count = new Array(n).fill(0)

  for (const [i, j] of pairs) {
    const dx = pos[j][0] - pos[i][0]
    const dy = pos[j][1] - pos[i][1]
    const d = Math.hypot(dx, dy)
    const stiffness = stiffnessOf(i, j, d)
    if (stiffness === null) continue

    const dSafe = d < 1e-9 ? 1e-9 : d
    const dirX = dx / dSafe
    const dirY = dy / dSafe
    const magnitude = (d - target) * stiffness
    const cx = dirX * magnitude
    const cy = dirY * magnitude

    sumX[i] += cx
    sumY[i] += cy
    count[i]++
    sumX[j] -= cx
    sumY[j] -= cy
    count[j]++
  }

  const correction: Vec2[] = Array.from({ length: n }, () => [0, 0] as Vec2)
  for (let i = 0; i < n; i++) {
    if (!freeMask[i] || count[i] === 0) continue
    correction[i][0] = sumX[i] / count[i]
    correction[i][1] = sumY[i] / count[i]
  }
  return correction
}

function maxNorm(correction: Vec2[]): number {
  let max = 0
  for (const [x, y] of correction) max = Math.max(max, Math.hypot(x, y))
  return max
}

function applyCorrection(pos: Vec2[], correction: Vec2[]): Vec2[] {
  return pos.map(
    ([x, y], i) => [x + correction[i][0], y + correction[i][1]] as Vec2
  )
}

/**
 * Yields the in-progress positions after every Jacobi correction (the
 * settle pass, then the collision-cleanup pass); the generator's return
 * value is the fully-converged result, same as physics.py's relax().
 */
export function* relaxSteps(
  positions: Vec2[],
  pinned: Set<number>,
  idVectors: number[][],
  config: PhysicsConfig
): Generator<Vec2[], Vec2[], void> {
  let pos: Vec2[] = positions.map(([x, y]) => [x, y])
  const n = pos.length
  if (n < 2) return pos

  const { contactDistance, tensionRange, tensionStiffness, settleStiffness, collisionEnabled } = config
  const freeMask = pos.map((_, i) => !pinned.has(i))
  const compatibility = tensionCompatibility(idVectors)

  for (let iter = 0; iter < config.settleIterations; iter++) {
    const pairs = findPairs(pos, tensionRange)
    const correction = pairCorrections(
      n,
      pos,
      pairs,
      contactDistance,
      (i, j, d) => {
        if (collisionEnabled && d < contactDistance) return settleStiffness
        if (d <= tensionRange) return tensionStiffness * compatibility[i][j]
        return null
      },
      freeMask
    )
    pos = applyCorrection(pos, correction)
    yield pos
    if (maxNorm(correction) < config.settleConvergenceTol) break
  }

  if (collisionEnabled) {
    for (let iter = 0; iter < config.cleanupIterations; iter++) {
      const pairs = findPairs(pos, contactDistance)
      const correction = pairCorrections(
        n,
        pos,
        pairs,
        contactDistance,
        (_i, _j, d) => {
          return d < contactDistance ? 1.0 : null
        },
        freeMask
      )
      pos = applyCorrection(pos, correction)
      yield pos
      if (maxNorm(correction) < config.cleanupConvergenceTol) break
    }
  }

  return pos
}

/** Drains relaxSteps() to convergence in one shot, for callers that don't need the animation. */
export function relax(
  positions: Vec2[],
  pinned: Set<number>,
  idVectors: number[][],
  config: PhysicsConfig
): Vec2[] {
  const gen = relaxSteps(positions, pinned, idVectors, config)
  let result = gen.next()
  while (!result.done) result = gen.next()
  return result.value
}

// Collision-cleanup budget for a single relaxTick() call — small and
// fixed rather than convergence-capped, since (unlike relaxSteps(),
// which runs once per discrete growth step and is expected to fully
// settle before the next one) this runs every animation frame forever;
// any overlap not fully resolved in one tick just keeps resolving over
// the next few frames instead, invisible at animation speed. Purely a
// client-side animation-pacing choice with no backend counterpart
// (unlike everything in PhysicsConfig), so it stays a local constant
// here rather than something sent over the wire. See
// sim/useRealtimeSimulation.ts.
const TICK_COLLISION_ITERATIONS = 4

/**
 * One frame's worth of physics: a single settle-pass correction
 * (tension + soft collision) followed by a small fixed number of hard
 * collision-cleanup corrections — never run to convergence the way
 * relax()/relaxSteps() are. Built for a caller that ticks this every
 * animation frame indefinitely (realtime growth) rather than one that
 * needs "fully settled" before moving on (the discrete step-based
 * replay). Reuses relaxSteps()'s exact same pair-finding/correction
 * math, just called once instead of looped to a tolerance.
 */
export function relaxTick(
  positions: Vec2[],
  pinned: Set<number>,
  idVectors: number[][],
  config: PhysicsConfig
): Vec2[] {
  let pos: Vec2[] = positions.map(([x, y]) => [x, y])
  const n = pos.length
  if (n < 2) return pos

  const { contactDistance, tensionRange, tensionStiffness, settleStiffness, collisionEnabled } = config
  const freeMask = pos.map((_, i) => !pinned.has(i))
  const compatibility = tensionCompatibility(idVectors)

  {
    const pairs = findPairs(pos, tensionRange)
    const correction = pairCorrections(
      n,
      pos,
      pairs,
      contactDistance,
      (i, j, d) => {
        if (collisionEnabled && d < contactDistance) return settleStiffness
        if (d <= tensionRange) return tensionStiffness * compatibility[i][j]
        return null
      },
      freeMask
    )
    pos = applyCorrection(pos, correction)
  }

  if (collisionEnabled) {
    for (let iter = 0; iter < TICK_COLLISION_ITERATIONS; iter++) {
      const pairs = findPairs(pos, contactDistance)
      const correction = pairCorrections(
        n,
        pos,
        pairs,
        contactDistance,
        (_i, _j, d) => {
          return d < contactDistance ? 1.0 : null
        },
        freeMask
      )
      pos = applyCorrection(pos, correction)
    }
  }

  return pos
}

File: trainer/backend/physics.py
"""Particle-based physics: nodes are circles of fixed radius. Two forces,
both purely distance-based and recomputed fresh every relax call — there
is no fixed topology, no notion of "this pair used to be connected":

- Surface tension (soft, breakable, short-range): any pair within
  TENSION_RANGE but not already touching gets pulled toward contact
  distance, only partially each iteration (TENSION_STIFFNESS < 1, so it
  behaves elastically). Past TENSION_RANGE there is no force at all —
  that absence is what makes it breakable. TENSION_RANGE is deliberately
  close to contact distance: making cohesion long-range turns it into
  "every node attracts every other node," which is an infeasible packing
  problem once more than a handful of circles are involved (in 2D only
  ~6 circles can mutually touch a shared neighbor) and the solver never
  settles.
- Collision (hard, unconditional): circles must never overlap. Run as a
  *separate* cleanup pass after tension has done its (soft) work, rather
  than mixed into the same pass — mixing a hard inequality constraint
  with a competing soft attraction is exactly the infeasible-packing
  problem above in miniature, one pair at a time, and doesn't reliably
  converge. Collision-only resolution has no competing force pulling
  things back together, so it always has a valid solution and converges
  reliably given enough iterations. Collision is purely geometric and
  never modulated by chemistry — circles can't overlap regardless of how
  compatible they are.

Tension strength is further modulated per-pair by the cosine similarity
of the two nodes' `id` vectors (see cell_state.py), clipped to
[0, 1] — homophilic-adhesion-style: similarly "identified" nodes pull
together at full strength, orthogonal or opposed identities feel no pull
at all even while in range, rather than negative/repulsive tension.
Proximity (TENSION_RANGE) decides whether a pair is *eligible* to feel
tension; compatibility decides how strongly they actually do.

Both passes use Jacobi-style updates (every node's correction this
iteration is the *average* of all its simultaneously-active pairwise
corrections, computed from one consistent snapshot of positions) rather
than Gauss-Seidel (correcting pairs one at a time, each seeing the
previous pair's already-updated positions) — Jacobi is what keeps this
stable at higher local density; Gauss-Seidel here visibly diverges once
more than a few circles crowd the same region.

Performance: TENSION_RANGE is a small, fixed interaction radius, so in
any reasonably spread-out graph, most pairs of nodes are never within
range of each other at all — the original implementation still checked
every one of them, every iteration, by building a dense (N, N) distance
matrix (O(n^2) per iteration, ~500 iterations per relax() call). This
version finds only the pairs actually within range via a KD-tree
(scipy.spatial.cKDTree.query_pairs — already a dependency, used
elsewhere for Chamfer distance) and does the correction math over just
those, which is where essentially all the wall-clock time was going
once graphs reached a couple hundred nodes.
"""

from __future__ import annotations

import numpy as np
from scipy.spatial import cKDTree

RADIUS = 0.5
CONTACT_DISTANCE = 2 * RADIUS
TENSION_RANGE = CONTACT_DISTANCE * 1.15
TENSION_STIFFNESS = 0.3
SETTLE_STIFFNESS = 0.5  # collision stiffness *within* the settle pass
SETTLE_ITERATIONS = 100
CLEANUP_ITERATIONS = 400

# Two separate tolerances, not one shared CONVERGENCE_TOL, because the two
# passes have very different tolerance-for-speed tradeoffs. Measured
# directly against a graph grown the normal way (single split + relax()
# per step, repeated 250x — trainer/backend's own real usage pattern, not
# a synthetic dense pile):
#
# - Settle (tension) is a *soft* constraint — its own docstring already
#   says it settles "only partially each iteration." A tight tolerance
#   here bought nothing: at 1e-4 the pass almost always burned its full
#   SETTLE_ITERATIONS cap without ever getting close (in the 250-split
#   benchmark, mean settle iterations *ran* was ~97 out of the 100 cap).
#   Loosening to 5e-3 cut settle iterations by ~76% (100 -> ~12-23 typical)
#   and total relax() iterations by ~40%, with node-to-node displacement
#   differences too small to see (well under 1% of a node's diameter in
#   the overwhelming majority of pairs) and — critically — no change in
#   remaining hard overlaps, since that's cleanup's job, not settle's.
# - Cleanup (collision) is the *hard* "circles never overlap" guarantee
#   (see this module's own docstring). Its own tolerance is much more
#   sensitive: loosening it from 1e-4 to 1e-3 in the same benchmark
#   roughly quintupled the count of still-overlapping pairs left behind
#   (real, visible circle interpenetration, not a cosmetic nicety), for
#   comparatively little iteration savings — cleanup's iteration count is
#   mostly driven by how many pairs actually need separating, not by how
#   tight its tolerance is. Left untouched.
SETTLE_CONVERGENCE_TOL = 5e-3
CLEANUP_CONVERGENCE_TOL = 1e-4

# target.py scales pixel targets so 1 pixel spacing = 1 unit of graph
# space; keep that convention meaningful without target.py needing to
# know anything about the particle model.
REST_LENGTH = CONTACT_DISTANCE


def _tension_compatibility(id_vectors: np.ndarray) -> np.ndarray:
    """(N, N) pairwise cosine similarity of id vectors, clipped to [0, 1]
    — 1 where two nodes' identities point the same way, 0 for orthogonal
    or opposed identities. Identity doesn't change during a relax call,
    so this is computed once and reused across iterations. Dense and
    O(n^2), unlike the per-iteration pair search below — but it only
    runs once per relax() call (not once per iteration), and a single
    matrix multiply over a few hundred nodes is microseconds; the actual
    per-iteration cost was always the position/distance sweep, not this.
    """
    norms = np.linalg.norm(id_vectors, axis=1, keepdims=True)
    norms_safe = np.where(norms < 1e-9, 1.0, norms)
    unit = id_vectors / norms_safe
    cosine_similarity = unit @ unit.T
    return np.clip(cosine_similarity, 0.0, 1.0)


def _pair_corrections(
    pos: np.ndarray,
    pairs: np.ndarray,
    target: np.ndarray,
    stiffness: np.ndarray,
    free_mask: np.ndarray,
) -> np.ndarray:
    """Jacobi-averaged correction from a sparse set of active pairs.
    `pairs` is (K, 2) int indices; `target`/`stiffness` are (K,) — one
    rest-distance and one stiffness value per pair, since settle mixes
    two different constraint types (collision vs. tension) in one pass
    and cleanup uses only one. Every pair contributes symmetrically to
    both endpoints (equal and opposite, mirroring the dense version's
    contrib[j,i] = -contrib[i,j]) via scatter-add, then each node's total
    is divided by how many active pairs it actually had — exactly the
    same Jacobi-average semantics as before, just computed over only the
    pairs that matter instead of every possible pair."""
    n = pos.shape[0]
    correction = np.zeros((n, 2))
    if pairs.shape[0] == 0:
        return correction

    i, j = pairs[:, 0], pairs[:, 1]
    diff = pos[j] - pos[i]
    dist = np.linalg.norm(diff, axis=1)
    dist_safe = np.where(dist < 1e-9, 1.0, dist)
    direction = diff / dist_safe[:, None]

    magnitude = (dist - target) * stiffness
    contrib = direction * magnitude[:, None]

    count = np.zeros(n)
    np.add.at(correction, i, contrib)
    np.add.at(correction, j, -contrib)
    np.add.at(count, i, 1.0)
    np.add.at(count, j, 1.0)

    count_safe = np.where(count == 0, 1.0, count)
    correction = correction / count_safe[:, None]
    correction[~free_mask] = 0.0
    return correction


def _resolve_collisions(pos: np.ndarray, free_mask: np.ndarray) -> np.ndarray:
    n = pos.shape[0]
    for _ in range(CLEANUP_ITERATIONS):
        pairs = cKDTree(pos).query_pairs(r=CONTACT_DISTANCE, output_type="ndarray")
        if pairs.shape[0] == 0:
            break

        target = np.full(pairs.shape[0], CONTACT_DISTANCE)
        stiffness = np.ones(pairs.shape[0])
        correction = _pair_corrections(pos, pairs, target, stiffness, free_mask)
        pos = pos + correction

        max_delta = float(np.max(np.linalg.norm(correction, axis=-1))) if n > 0 else 0.0
        if max_delta < CLEANUP_CONVERGENCE_TOL:
            break
    return pos


def relax(positions: np.ndarray, pinned: set[int], id_vectors: np.ndarray) -> np.ndarray:
    pos = positions.copy()
    n = pos.shape[0]
    if n < 2:
        return pos

    free_mask = np.ones(n, dtype=bool)
    for idx in pinned:
        if 0 <= idx < n:
            free_mask[idx] = False

    compatibility = _tension_compatibility(id_vectors)

    for _ in range(SETTLE_ITERATIONS):
        pairs = cKDTree(pos).query_pairs(r=TENSION_RANGE, output_type="ndarray")
        if pairs.shape[0] == 0:
            break

        i, j = pairs[:, 0], pairs[:, 1]
        dist = np.linalg.norm(pos[j] - pos[i], axis=1)
        collision = dist < CONTACT_DISTANCE
        stiffness = np.where(collision, SETTLE_STIFFNESS, TENSION_STIFFNESS * compatibility[i, j])
        target = np.full(pairs.shape[0], CONTACT_DISTANCE)

        correction = _pair_corrections(pos, pairs, target, stiffness, free_mask)
        pos = pos + correction

        max_delta = float(np.max(np.linalg.norm(correction, axis=-1))) if n > 0 else 0.0
        if max_delta < SETTLE_CONVERGENCE_TOL:
            break

    # Hard guarantee, decoupled from the soft settling above: circles
    # never overlap, full stop.
    pos = _resolve_collisions(pos, free_mask)

    return pos
